Add Stelpro (Stello) Allia HT402 power and energy sensors

ouaibe opened the pull request:
## Proposed change
This is a quirks V2 configuration that allows reading of the proprietary/manufacturer power consumption (W) and Energy (Wh) attributes for Stelpro (identified as "Stello" by ZHA) Allia HT402 (Hilo-sold) thermostats.


## Additional information
These attributes (0x4009 and 0x4008) have been identified from https://github.com/Koenkk/zigbee2mqtt/issues/14651 (all works in Z2M flawlessly but not in ZHA).
The quirks makes them visible in HA as Power (immediate) and Energy (cumulative) sensors for every HT402 Thermostats connected.
<img width="346" height="310" alt="image" src="https://github.com/user-attachments/assets/1cd08f59-0463-4b24-a431-9346f63fe44e" />


## Device diagnostics

```
2025-10-26 08:57:48.048 DEBUG (MainThread) [zha.zigbee.cluster_handlers] [**REDACTED**]: cluster_handler[thermostat] attribute_updated - cluster[Thermostat] attr[16392] value[846]
2025-10-26 08:57:48.048 DEBUG (MainThread) [zha] Emitting event cluster_handler_attribute_updated with data ClusterAttributeUpdatedEvent(attribute_id=16392, attribute_name=16392, attribute_value=846, cluster_handler_unique_id='**REDACTED**', cluster_id=513, event_type='cluster_handler_event', event='cluster_handler_attribute_updated') (8 listeners)
2025-10-26 08:57:48.049 DEBUG (MainThread) [zha.zigbee.cluster_handlers] [**REDACTED**]: cluster_handler[thermostat] attribute_updated - cluster[Thermostat] attr[16393] value[59]
2025-10-26 08:57:48.049 DEBUG (MainThread) [zha] Emitting event cluster_handler_attribute_updated with data ClusterAttributeUpdatedEvent(attribute_id=16393, attribute_name=16393, attribute_value=59, cluster_handler_unique_id='**REDACTED**', cluster_id=513, event_type='cluster_handler_event', event='cluster_handler_attribute_updated') (8 listeners)
```

[zha-8fdbbba0cd6c61f53202f4f5a2f05d82-Stello HT402-708e916b645a299a90a9f1b6289e1505.json](https://github.com/user-attachments/files/23261633/zha-8fdbbba0cd6c61f53202f4f5a2f05d82-Stello.HT402-708e916b645a299a90a9f1b6289e1505.json)


## Checklist

- [X] The changes are tested and work correctly
- [X] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works
- [X] Device diagnostics data has been attached

ouaibe committed:
Create allia_ht402.py

ouaibe committed:
Update allia_ht402.py

Initial version that works in HA.

ouaibe committed:
Update allia_ht402.py

Adding few doc URLs.

ouaibe committed:
Create __init__.py

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

ouaibe committed:
Update allia_ht402.py

Adding docstrings to satisfy pydocstyle (pep257) requirements.

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

charlesp-l commented:
+1 to get this added! I know there is also a way to expose a field to [display the exterior temperature](https://www.zigbee2mqtt.io/devices/HT402.html#notes)

It seems trivial in the Zigbee2MQTT world, I don't know what it would look like in the ZHA world.
https://github.com/Koenkk/zigbee-herdsman-converters/blob/cfe8de84ca6f0d40c5a1cdf3f95fb3b6bdd5346e/src/devices/stelpro.ts#L290C1-L290C55

ouaibe commented:
Yes I actually have a local working version of that external temperature display - it's perfectible (as the sensor created on the HA devices is off by x100C because of the Stelpro format), but the display on the termostats themselves works fine.

It allows you to read from a temperature readout anywhere and update it accordingly on the termostats in your home.

I will commit these additional options to the PR later.

ouaibe committed:
Update allia_ht402.py

Add the outdoor temperature support with the caveat that it is off by x100C as it is the necessary format for the thermostat to display it properly and QuirksV2 don't seem to support inline lambda  to_value/from_value in ZCLAttributeDef.

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

ouaibe commented:
Using the code newly pushed to the PR, here's an HA automation you can use to update the external temperature every 15min:

- Reads the temperature from sensor.openweathermap_temperature
- Creates a list of `ieee` addresses for all your thermostats
- Pushes the new attribute to them using `zha.set_zigbee_cluster_attribute`

```
alias: Push outdoor temp to Allia thermostats (every 15 min, ZHA)
triggers:
  - minutes: /15
    trigger: time_pattern
conditions:
  - condition: template
    value_template: "{{ temp_c_x100 is not none }}"
  - condition: template
    value_template: "{{ thermostats | count > 0 }}"
actions:
  - repeat:
      for_each: "{{ thermostats }}"
      sequence:
        - data:
            ieee: "{{ repeat.item }}"
            endpoint_id: 25
            cluster_id: 513
            cluster_type: in
            attribute: 16385
            value: "{{ temp_c_x100 }}"
          action: zha.set_zigbee_cluster_attribute
variables:
  temp_c: "{{ states('sensor.openweathermap_temperature') | float(0) }}"
  temp_c_x100: "{{ (temp_c * 100) | round(0) | int }}"
  thermostats: >
    {% set pattern = '^sensor\.thermostat_.*_outdoor_temperature$' %} {% set out
    = namespace(list=[]) %} {% for e in states.sensor |
    selectattr('entity_id','match',pattern) %}
      {% set did = device_id(e.entity_id) %}
      {% set ieee = (device_attr(did, 'identifiers')
                     | selectattr(0, 'equalto', 'zha')
                     | map('last') | list | first) %}
      {% if ieee %}{% set out.list = out.list + [ieee] %}{% endif %}
    {% endfor %} {{ out.list }}
```

Proof that it works:
----
<img width="342" height="60" alt="image" src="https://github.com/user-attachments/assets/2a702f3c-84ec-4d1f-a2fc-478740708cbb" />

<img width="793" height="766" alt="image" src="https://github.com/user-attachments/assets/269c39b3-066a-4233-938e-58018a5f5557" />

charlesp-l commented:
It (mostly) works on my installation. Power and energy seem to work as expected! The outdoor temperature does have some issues...  
1. The writable outdoor temperature isn't shown as `outdoor_temperature`, but as `sensor.basement_thermostat_temperature`: 
<img width="642" height="1034" alt="image" src="https://github.com/user-attachments/assets/c41fa9be-bcde-4634-922b-fedad6d2480c" />
<img width="1092" height="814" alt="image" src="https://github.com/user-attachments/assets/db4be6de-9399-45a2-936e-e0ca2b2353e5" />
 (The device diagnostic does show the `allia_outdoor_temperature`, but the HA UI doesn't, not sure exactly why.)

2. As you can guess, the automation therefore fails as the `thermostats` template doesn't get any sensor that matches `'^sensor\.thermostat_.*_outdoor_temperature$'`.   
3. (Suggestion) Maybe the `outdoor_temperature` could be set as writable only, instead of readable/write, as it's not a sensor per se. I don't know if there something else than a `sensor` to describe this? But that's not very important. 

Thanks for your work!

lit-af commented:
Hi guys,

Would you guys be interested adding control for the Challenge In Progress Icon on the HT402?

To turn it on:
```yaml
action: zha_toolkit.attr_write
data:
  ieee: climate.stello_ht402
  endpoint_id: 25
  cluster: 0x0201
  attribute: 0x4105
  attr_val: 25200
  value_type: uint16
```

To turn it off:
```yaml
action: zha_toolkit.attr_write
data:
  ieee: climate.stello_ht402
  endpoint_id: 25
  cluster: 0x0201
  attribute: 0x4105
  attr_val: 0
  value_type: uint16
  ```

ouaibe commented:
@lit-af that's a good idea! Supporting all the proprietary flags/attributes of that thermostat is a great objective.

Feel free to:

- Confirm the values that can be set fo that proprietary attribute (25200 seems oddly specific?).
- Implement and test locally.
- Fork and propose additional code change.

marc3k18 commented:
This is release or not yet?

lit-af commented:
@ouaibe The values 25200 and 0 were found by sniffing ZigBee packets from the Hilo hub at the beginning/end of a challenge. 
I have a forked version running locally and the challenge icon works, but I have an issue with the thermostat current temperature reading for some reason I haven't debugged yet.

@marc3k18 No, this is still a WIP.

ouaibe committed:
Merge branch 'dev' into dev

lit-af commented:
So I modified the original quirk and it's working ok.

- I renammed the quirk to match the naming of z2m.
- The Peak Event Icon is now supported. 
- Outdoor temperature is now a number instead of a sensor. Based on: https://github.com/zigpy/zha-device-handlers/pull/4042/files#diff-17dd0e34db789810c0542b778efbd5589e903c35111f601ceed7bb52e1d5c8abR557
- I noticed an issue where after the first pairing the current room temperature isn't reported properly until the attribute is manually read. 
- There might be a way to add a function to handle the rounding and scaling for the outdoor temperature display.

```python
"""ZHA Quirk (v2) for Stello HT402 thermostat.

Adds manufacturer-specific attributes to the Thermostat cluster:
- 0x4001: Outdoor temperature reference (int16, write-only, value needs to be scaled x1000)
- 0x4008: Instant power (uint16, read-only)
- 0x4009: Cumulative energy (uint32, read-only)
- 0x4105: Peak demand event icon (uint16, write-only)
"""

from typing import Final
from enum import IntEnum

from zigpy.quirks import CustomCluster
from zigpy.quirks.v2 import (
    QuirkBuilder,
    ReportingConfig,
    SensorDeviceClass,
    SensorStateClass,
)
from zigpy.quirks.v2.homeassistant import (
    UnitOfEnergy,
    UnitOfPower
)
import zigpy.types as t
from zigpy.zcl.clusters.hvac import Thermostat
from zigpy.zcl.foundation import ZCLAttributeDef

# ──────────────────────────────────────────────────────────────
# Device identification
# ──────────────────────────────────────────────────────────────

MANUFACTURER: Final = "Stello"
MODEL: Final = "HT402"
EP_THERMOSTAT: Final = 25  # 0x19

# ──────────────────────────────────────────────────────────────
# Named values
# ──────────────────────────────────────────────────────────────

class PeakDemandEventIcon(IntEnum):
    OFF = 0x0000
    ON = 0x6270

# ──────────────────────────────────────────────────────────────
# Custom Thermostat cluster
# ──────────────────────────────────────────────────────────────

class StelloThermostatCluster(Thermostat, CustomCluster):
    """Thermostat cluster extended with Stello manufacturer attributes."""

    class AttributeDefs(Thermostat.AttributeDefs):
        # 0x4001: Outdoor temperature reference (scaled x1000), write-only
        stello_outdoor_temperature = ZCLAttributeDef(
            id=0x4001,
            type=t.int16s,
            access="w",
        )

        # 0x4008: Instant power (W), read-only
        stello_power_w = ZCLAttributeDef(
            id=0x4008,
            type=t.uint16_t,
            access="rp",
        )

        # 0x4009: Cumulative energy (Wh), read-only
        stello_energy_wh = ZCLAttributeDef(
            id=0x4009,
            type=t.uint32_t,
            access="rp",
        )

        # 0x4105: Peak demand event icon, write-only
        stello_peak_demand_event_icon = ZCLAttributeDef(
            id=0x4105,
            type=t.uint16_t,
            access="w",
        )

# ──────────────────────────────────────────────────────────────
# Quirk registration
# ──────────────────────────────────────────────────────────────

(
    QuirkBuilder(MANUFACTURER, MODEL)
    .replaces(StelloThermostatCluster, endpoint_id=EP_THERMOSTAT)

    # Outdoor temperature reference (scaled x1000)
    .number(
        attribute_name=StelloThermostatCluster.AttributeDefs
            .stello_outdoor_temperature.name,
        cluster_id=StelloThermostatCluster.cluster_id,
        endpoint_id=EP_THERMOSTAT,
        min_value=-32768,  # int16 minimum
        max_value=19000,   # int16 maximum
        step=100,         # Display max resolution
        unit="temperature",
        translation_key="stello_outdoor_temperature",
        fallback_name="Outdoor Temperature",
    )

    # Instant power
    .sensor(
        attribute_name=StelloThermostatCluster.AttributeDefs
            .stello_power_w.name,
        cluster_id=StelloThermostatCluster.cluster_id,
        endpoint_id=EP_THERMOSTAT,
        device_class=SensorDeviceClass.POWER,
        state_class=SensorStateClass.MEASUREMENT,
        unit=UnitOfPower.WATT,
        reporting_config=ReportingConfig(
            min_interval=5,
            max_interval=300,
            reportable_change=1,
        ),
        translation_key="stello_power_w",
        fallback_name="Power",
    )

    # Energy (cumulative)
    .sensor(
        attribute_name=StelloThermostatCluster.AttributeDefs
            .stello_energy_wh.name,
        cluster_id=StelloThermostatCluster.cluster_id,
        endpoint_id=EP_THERMOSTAT,
        device_class=SensorDeviceClass.ENERGY,
        state_class=SensorStateClass.TOTAL_INCREASING,
        unit=UnitOfEnergy.WATT_HOUR,
        reporting_config=ReportingConfig(
            min_interval=30,
            max_interval=3600,
            reportable_change=10,
        ),
        translation_key="stello_energy_wh",
        fallback_name="Energy",
    )

    # Peak demand event icon
    .number(
        attribute_name=StelloThermostatCluster.AttributeDefs
            .stello_peak_demand_event_icon.name,
        cluster_id=StelloThermostatCluster.cluster_id,
        endpoint_id=EP_THERMOSTAT,
        min_value=int(PeakDemandEventIcon.OFF),
        max_value=int(PeakDemandEventIcon.ON),
        step=int(PeakDemandEventIcon.ON),
        translation_key="stello_peak_demand_event_icon",
        fallback_name="Peak Demand Event",
    )

    .add_to_registry()
)
```

TheJulianJES commented on zhaquirks/stelpro/allia_ht402.py:
```diff
@@ -0,0 +1,94 @@
+        translation_key="allia_energy_wh",
+        fallback_name="Allia Energy",
+    )
+    # Outdoor Temperature (0x4001) — read/write
+    .sensor(
+        attribute_name=AlliaThermostatCluster.AttributeDefs.allia_outdoor_temperature.name,
+        cluster_id=AlliaThermostatCluster.cluster_id,
+        endpoint_id=EP_THERMOSTAT,
+        device_class=SensorDeviceClass.TEMPERATURE,
+        state_class=SensorStateClass.MEASUREMENT,
+        unit=UnitOfTemperature.CELSIUS,  # Would need to convert to F
+        translation_key="allia_outdoor_temperature",
+        fallback_name="Outdoor Temperature",
+    )
```
This is fine if we don't want to write to it.

TheJulianJES commented on zhaquirks/stelpro/allia_ht402.py:
```diff
@@ -0,0 +1,94 @@
+        endpoint_id=EP_THERMOSTAT,
+        device_class=SensorDeviceClass.TEMPERATURE,
+        state_class=SensorStateClass.MEASUREMENT,
+        unit=UnitOfTemperature.CELSIUS,  # Would need to convert to F
```
I'm assuming this was tested correctly, so this is fine.

TheJulianJES commented on zhaquirks/stelpro/allia_ht402.py:
```diff
@@ -0,0 +1,94 @@
+            min_interval=5, max_interval=300, reportable_change=1
+        ),
+        translation_key="allia_power_w",
+        fallback_name="Allia Power",
```
Why do we need to include "Allia" in the entity name? It should probably just be "Power" or "Energy", so I'd suggest removing the `translation_key` to have HA use the translated variant of the device class name.
```suggestion
        fallback_name="Power",
```
The `fallback_name` won't actually be used then, but it's just here to represent what HA likely uses.

---

Also, per HA naming conventions, we need to use sentence case, so this should have been "Allia power", so a lower-case "p". But as we shouldn't include that anyway, it doesn't matter here.

TheJulianJES commented on zhaquirks/stelpro/allia_ht402.py:
```diff
@@ -0,0 +1,94 @@
+            min_interval=30, max_interval=3600, reportable_change=10
+        ),
+        translation_key="allia_energy_wh",
+        fallback_name="Allia Energy",
```
Same applies here as well:
```suggestion
        fallback_name="Energy",
```

TheJulianJES commented on zhaquirks/stelpro/allia_ht402.py:
```diff
@@ -0,0 +1,94 @@
+        state_class=SensorStateClass.MEASUREMENT,
+        unit=UnitOfTemperature.CELSIUS,  # Would need to convert to F
+        translation_key="allia_outdoor_temperature",
+        fallback_name="Outdoor Temperature",
```
[`outdoor_temperature`](https://github.com/home-assistant/core/blob/10d32b7f23b0a399e845ea9f047ed3eb99b23864/homeassistant/components/zha/strings.json#L791-L793) already exists for a translation, so we should use that. If we kept `allia_outdoor_temperature`, it would have meant more translation work.

Also, sentence case. So:
```suggestion
        translation_key="outdoor_temperature",
        fallback_name="Outdoor temperature",
```

TheJulianJES commented on zhaquirks/stelpro/allia_ht402.py:
```diff
@@ -0,0 +1,94 @@
+from zigpy.zcl.clusters.hvac import Thermostat
+from zigpy.zcl.foundation import ZCLAttributeDef
+
+MANUFACTURER: Final = "Stello"
+MODEL: Final = "HT402"
```
Let's explicitly use them in `QuirkBuilder` below. No need for a constant here.

TheJulianJES commented on zhaquirks/stelpro/allia_ht402.py:
```diff
@@ -0,0 +1,94 @@
+
+
+(
+    QuirkBuilder(MANUFACTURER, MODEL)
```
So, use here:
```suggestion
    QuirkBuilder("Stello", "HT402")
```

TheJulianJES commented on zhaquirks/stelpro/allia_ht402.py:
```diff
@@ -0,0 +1,94 @@
+
+MANUFACTURER: Final = "Stello"
+MODEL: Final = "HT402"
+EP_THERMOSTAT: Final = 25  # 0x19
```
I think we also don't need to define this as a constant. I'd rather it be used explicitly everywhere it's needed for quirks v2 entities.

ouaibe replied:
Resolved in latest version.

TheJulianJES commented on zhaquirks/stelpro/allia_ht402.py:
```diff
@@ -0,0 +1,94 @@
+        # 0x4008: Instant power in Watts (uint16)
+        allia_power_w = ZCLAttributeDef(id=0x4008, type=t.uint16_t, access="rp")
+        # 0x4009: Cumulative energy in Watt-hours (uint32)
+        allia_energy_wh = ZCLAttributeDef(id=0x4009, type=t.uint32_t, access="rp")
```
I don't think we should prefix these with the device name either. `outdoor_temperature`, `power`, and `energy` should be enough IMO.

ouaibe replied:
Resolved in latest version.

ouaibe replied:
Thanks for the exhaustive review, really appreciated!

TheJulianJES reviewed (commented):
Thanks for the PR. Left some feedback

ouaibe committed:
Update zhaquirks/stelpro/allia_ht402.py

Co-authored-by: TheJulianJES <TheJulianJES@users.noreply.github.com>

ouaibe committed:
Update zhaquirks/stelpro/allia_ht402.py

Co-authored-by: TheJulianJES <TheJulianJES@users.noreply.github.com>

ouaibe committed:
Update zhaquirks/stelpro/allia_ht402.py

Co-authored-by: TheJulianJES <TheJulianJES@users.noreply.github.com>

ouaibe committed:
Update zhaquirks/stelpro/allia_ht402.py

Co-authored-by: TheJulianJES <TheJulianJES@users.noreply.github.com>

ouaibe committed:
Merge branch 'dev' into dev

ouaibe committed:
Update allia_ht402.py per code review.

Remove all "allia_" prefixes.
Remove EP_THERMOSTAT final variable and use in-line value `25` instead where needed.

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

ouaibe commented:
> So I modified the original quirk and it's working ok.
> 
>     * I renammed the quirk to match the naming of z2m.
> 
>     * The Peak Event Icon is now supported.
> 
>     * Outdoor temperature is now a number instead of a sensor. Based on: https://github.com/zigpy/zha-device-handlers/pull/4042/files#diff-17dd0e34db789810c0542b778efbd5589e903c35111f601ceed7bb52e1d5c8abR557
> 
>     * I noticed an issue where after the first pairing the current room temperature isn't reported properly until the attribute is manually read.
> 
>     * There might be a way to add a function to handle the rounding and scaling for the outdoor temperature display.
> 
> 

These are great elements, thanks for the hard work. Here's what I suggest @lit-af:

- Let's finish the work on a V1 that passes all the QuirksV2 standard coding practice based on the original PR being reviewed right now by the ZHA team.
- Once it's been accepted as part of zha-quirks, we can look into backporting your additions to it: the challenge ON/OFF element and seeing whether changing the outdoor_temp to a `number` makes sense.

ouaibe committed:
Merge branch 'dev' into dev

ouaibe committed:
Merge branch 'dev' into dev

lit-af commented:
Have a look at the Z2M documentation for a solution to the energy value stuck at 500 https://www.zigbee2mqtt.io/devices/HT402.html#reseting-the-energy-value

jfsimoneau commented:
> Have a look at the Z2M documentation for a solution to the energy value stuck at 500 https://www.zigbee2mqtt.io/devices/HT402.html#reseting-the-energy-value

In my setup, I use `zha_toolkit.misc_settime` every hour on my devices and it works.

jfsimoneau commented:
@lit-af Looking at z2m code, the peak icon values are actually the number of seconds to display the icon. I checked this locally and writing 10 makes the icon appear only for 10 seconds. I will try to replicate that locally and provide a PR to @ouaibe.

lit-af commented:
Yes, that's it! The value is the amount of time the icon is displayed in seconds.
Have a look at the code in my comment: https://github.com/zigpy/zha-device-handlers/pull/4453#issuecomment-3719146973

jfsimoneau commented:
@lit-af If you want to take a look, I have a PR (https://github.com/ouaibe/zha-device-handlers/pull/1) for those features. I use that quirk locally.

lit-af commented:
I switched to z2m a while back, but your PR seems good :+1:

jfsimoneau reviewed (approved):
I based my custom quirk out of this, so I can make a PR with the new functionalities when this is merged.
